### PR TITLE
Update how_to_convert_image_to_ofrecord.md

### PR DESCRIPTION
**Pull request by Ziheng-ZHAO**

### CODE REVIEW

File: en/docs/extended_topics/consistent_mirrored.md [hunk 1]
```diff
@@ -40,36 +39,36 @@ The first matrix in grey on the left-hand side of the equation is the input samp
 In this section, we will see how the operators above are split in different ways in data parallelism and model parallelism.
 
 
-### Data parallelism diagram
+### Data Parallelism Diagram
 
 In **data parallelism**, the sample data are divided into small parts. **The divided data** will send to each training nodes and calculate with the **complete models**. Finally, we combine the information in each node. As shown in the figure below:
```
Comment (MARD1NO) [inline en/docs/extended_topics/consistent_mirrored.md:45]: each training nodes -> each training node

File: en/docs/extended_topics/consistent_mirrored.md [hunk 2]
```diff
@@ -40,36 +39,36 @@ The first matrix in grey on the left-hand side of the equation is the input samp
...
 We will introduce two parallelism strategies in OneFlow (`mirrored` and `consistent`) and learn how to choose different parallelism methods in different strategies.
 
-### Two types of Placeholder
-In [use OneFlow build neural network](../basics_topics/build_nn_with_op_and_layer.md), we have already introduced the concept of  `Placeholder` and `Blob`. 
+### Two Types of Placeholder
+In [use OneFlow build neural network](../basics_topics/build_nn_with_op_and_layer.md) and [The Definition and Call of Job Function](. /job_function_define_call.md), we have already introduced the concept of  `Placeholder` and `Blob`. 
 
 Actually, in the view of parallelism, the  `Placeholder` of OneFlow can be divided into two types: Use `oneflow.typing.Numpy.Placeholder` and `oneflow.typing.ListNumpy.Placeholder` to construct the placeholder, which is corresponding to `Consistent`  and `Mirrored`.
 
 We will explain them in detail in the examples below.
```
Comment (MARD1NO) [inline en/docs/extended_topics/consistent_mirrored.md:68]: 百度了下 in detail 表示详细地， in details 指各个细节。我感觉这里两者都可以，改不改可以自己斟酌下

File: en/docs/extended_topics/consistent_mirrored.md [hunk 3]
```diff
@@ -86,77 +85,14 @@ The return value type of job function is `oneflow.typing.ListNumpy`. Every eleme
...
 **Concat** all elements in the `list` can make a complete batch.
 
-### Code Example
+### Code 
 
 In the following code, we use `mirrored_view` with two devices to train.
 
 Complete Code: [mirrored_strategy.py](../code/extended_topics/mirrored_strategy.py)
 
 We will explain the key part of the code in detail in the following "code explanation" later section
```
Comment (MARD1NO) [inline en/docs/extended_topics/consistent_mirrored.md:95]: following 和 later 感觉有点重复

File: en/docs/extended_topics/debug_by_vscode.md [hunk 4]
```diff
@@ -55,9 +53,7 @@ After some simple configuration, clangd can provide us with code completion, sym
 Followings are required before we configure clangd:
 
 * We have already compiled OneFlow and generated `compile_commands.json` file.
-
 * We have already installed clangd on remote host through "Remote - SSH".
-
 * It is **NOT** recommended to install the extension "ms-vscode.cpptools C/C++" which is recommended by VS Code. Because it conflicts with clangd.
```
Comment (MARD1NO) [inline en/docs/extended_topics/debug_by_vscode.md:58]: it is not recommended to do 这句语法是错误的
可以改成 it is recommended that xxxxx，后面的自己变一下

File: en/docs/extended_topics/how_to_convert_image_to_ofrecord.md [hunk 5]
```diff
@@ -141,5 +141,4 @@ The outputs of training script should like below:
 
 At this point, we have successfully completed the whole process of data set production, reading and training.
```
Comment (MARD1NO) [inline en/docs/extended_topics/how_to_convert_image_to_ofrecord.md:143]: data set -> dataset

File: en/docs/extended_topics/how_to_make_ofdataset.md [hunk 6]
```diff
@@ -1,6 +1,5 @@
-## Loading and Preparing OFRecord Dataset
 
-In article [data_input](../basics_topics/data_input.md), we know that multi-threading, scheduling of resources and other supporting process increase the efficiency of processing data in OneFlow data-pipeline. Also, we learn about the operation in data-pipeline.
+In [data input] (... /basics_topics/data_input.md) we learned that it is usually more efficient to load data using DataLoader and related operators. We also learned how to use DataLoader and related operators.
```
Comment (MARD1NO) [inline en/docs/extended_topics/how_to_make_ofdataset.md:3]: 我觉得把后面那句 we also 拆成 Also，we xxx 会好一点

File: en/docs/extended_topics/how_to_make_ofdataset.md [hunk 7]
```diff
@@ -12,7 +11,7 @@ In this article, we will focus on the loading and generating of OneFlow's OFReco
 
 * The transition between OFRecord dataset and other data formats
 
-## What is OFRecord dataset
+## What is OFRecord Dataset
 
 In article [OFRecord](ofrecord.md), we introduce what `OFRecord file ` is and the storage format of `OFRecord file`.
```
Comment (MARD1NO) [inline en/docs/extended_topics/how_to_make_ofdataset.md:17]: `OFRecord file `    file 后面多了个空格

File: en/docs/extended_topics/how_to_make_ofdataset.md [hunk 8]
```diff
@@ -70,60 +69,13 @@ Actually, we can specify the filename prefix `part-`, whether we pad the filenam
 
 OneFlow provides the API interface to load OFRecord dataset by specifying the path of dataset directory, so that we can have the multi-threading, pipelining and some other advantages brought by OneFlow framework.
 
-## The method to load OFRecord dataset
+## The Method to Load OFRecord Dataset
 
 We use `ofrecord_reader` to load and preprocess dataset. 
 
 In article [Data Input](../basics_topics/data_input.md), we show how to use `ofrecord_reader` API to load OFRecord data and preprocess it.
```
Comment (MARD1NO) [inline en/docs/extended_topics/how_to_make_ofdataset.md:77]: we show -> we have shown

File: en/docs/extended_topics/how_to_make_ofdataset.md [hunk 9]
```diff
@@ -150,12 +102,10 @@ def ofrecord_reader(
 - `shuffle_buffer_size` assign the buffer size when reading data
 - `shuffle_after_epoch` assign whether shuffle the sample order after each epoch
 
-The advantage of using `ofrecord_reader` is that data processing in `ofrecord_reader` is scheduled by OneFlow and enjoys the OneFlow pipelining speedup.  
-
-For flexibility and extensibility of the code, we can define a preprocessing OP for `ofrecord_reader` to deal with specific data formats which are coupled with business logic.
-
-- Please refer to [Data input](../basics_topics/data_input.md) to learn about data pipelining and preprocess
-- Please refer to [User op](./user_op.md) to learn about customized OP
+The benefit of using `ofrecord_reader` is that `ofrecord_reader` acts as a normal operator which participates in OneFlow composition optimization and enjoys OneFlow flowline acceleration.
```
Comment (MARD1NO) [inline en/docs/extended_topics/how_to_make_ofdataset.md:105]: flowline 改成 pipeline

File: en/docs/extended_topics/how_to_make_ofdataset.md [hunk 10]
```diff
@@ -150,12 +102,10 @@ def ofrecord_reader(
...
+For flexibility and extensibility of the code, we can define a preprocessing OP for `ofrecord_reader` to deal with specific data formats which are coupled with operational logic (e.g. decoding, decompression and etc.).
+- For more information on DataLoader and related operator usage refer to [Data input](../basics_topics/data_input.md) .
+-  For more information on customized OP refer to [User op](./user_op.md).
```
Comment (MARD1NO) [inline en/docs/extended_topics/how_to_make_ofdataset.md:109]: 108行这里“-” 后面多了个空格。

另外refer前面加点词比较好，比如you can refer to xxx, please refer to xxx

File: en/docs/extended_topics/model_mixed_parallel.md [hunk 11]
```diff
@@ -32,11 +32,11 @@ Compare the figure above, we can easily get the logic of the network:
...
 In `consistent` view, OneFlow supports the data parallelism, model parallelism and hybrid parallelism. We will introduce them in order but hybrid parallelism is the key point.
 
-## The features of parallelism in consistent view
+## The Features of Parallelism in Consistent View
 
-### Data parallelism
+### Data Parallelism
 
-We have already known that in consistent view. The default parallelism method is data parallelism. If we choose mirrored view, we can only use data parallelism. If numpy data is passed directly when the job function is called (instead of using OneFlow's `flow.data.xxx_reader` interface for data loading), the difference is that: 
+We have already known that in consistent view. The default parallelism method is data parallelism. If we choose mirrored view, we can only use data parallelism. If you pass `numpy` data directly when you call the job function (instead of using OneFlow's [DataLoader and related operators] (... /basics_topics/data_input.md#dataloader)), the difference between the two are:
```
Comment (MARD1NO) [inline en/docs/extended_topics/model_mixed_parallel.md:40]: between the two 改成 between them 或者 between the two views are

File: en/docs/extended_topics/model_mixed_parallel.md [hunk 12]
```diff
@@ -187,106 +130,22 @@ In addition, `flow.layers.dense`  use `model_distribute`  to set parallelism mod
 
 As you can see, we can change the single machine training program to a distributed, hybrid parallel program with few modifications, which is one of the features that distinguishes OneFlow from other frameworks.
 
-## Pipelining example
+## Pipelining Example
 
 Besides the model parallelism, OneFlow also provides a more flexible parallelism method called pipelining, it allow user use  `scope.placement` to specify the device of the operator.
```
Comment (MARD1NO) [inline en/docs/extended_topics/model_mixed_parallel.md:135]: use 后面多了个空格

File: en/docs/basics_topics/distributed_train.md [hunk 13]
```diff
@@ -109,74 +109,38 @@ def config_distributed():
...
-        for i, (images, labels) in enumerate(zip(train_images, train_labels)):
-            loss = train_job(images, labels)
-            if i % 20 == 0:
-                print(loss.mean())
+> Make sure the version of the communication library (nccl) is the same on each machine during distributed training.
 
-```
+- Using virtual network cards
 
+> If there is a virtual network cards, you may not be able to communicate with nccl. In this case, you need to specify the communication network cards by `export NCCL_SOCKET_IFNAME=device_name`. More details please refer to [nccl official documentation](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs).
```
Comment (MARD1NO) [inline en/docs/basics_topics/distributed_train.md:146]: there is a -> there are 因为你后面是cards

File: en/docs/basics_topics/model_load_save.md [hunk 14]
```diff
@@ -74,36 +74,44 @@ The following example use `get_variable` to create parameters and build the netw
...
 
+* [xavier_normal_initializer](https://oneflow.readthedocs.io/en/master/oneflow.html#oneflow.xavier_normal_initializer)
 
-## The python interface of OneFlow models
+* [xavier_uniform_initializer](https://oneflow.readthedocs.io/en/master/oneflow.html#oneflow.xavier_uniform_initializer)
+
+
+
+
+
```
Comment (MARD1NO) [inline en/docs/basics_topics/model_load_save.md:113]: 这么多换行是为什么

File: en/docs/basics_topics/model_load_save.md [hunk 15]
```diff
@@ -339,105 +347,26 @@ WARNING! CANNOT find variable path in : ./mlp_models_1/dense3-weight/out. It wil
...
-            dense2, 10, kernel_initializer=initializer, name="dense3"
-        )
-        loss = flow.nn.sparse_softmax_cross_entropy_with_logits(labels, dense3)
+When the training is complete, you will get the `mlp_models_1` in the current working directory.
 
-    lr_scheduler = flow.optimizer.PiecewiseConstantScheduler([], [0.1])
-    flow.optimizer.SGD(lr_scheduler, momentum=0).minimize(loss)
-
-    return loss
+The following code is from [mlp_mnist_finetune.py](../code/basics_topics/mlp_mnist_finetune.py). After finetuning (add one more layer `dense3` in backbone network), we load  `./mlp_models_1` and train it.
```
Comment (MARD1NO) [inline en/docs/basics_topics/model_load_save.md:363]: load 后面多了空格

File: en/docs/basics_topics/model_load_save.md [hunk 16]
```diff
@@ -339,105 +347,26 @@ WARNING! CANNOT find variable path in : ./mlp_models_1/dense3-weight/out. It wil
...
-        loss = train_job(images, labels)
-        if i % 20 == 0:
-            print(loss.mean())
-    check_point.save("./mlp_ext_models_1")
+```shell
+wget https://docs.oneflow.org/code/basics_topics/mlp_mnist_finetune.py
+python3 mlp_mnist_finetune.py
 ```
+
+The finely tuned models are saved in `. /mlp_ext_models_1`.
```
Comment (MARD1NO) [inline en/docs/basics_topics/model_load_save.md:372]: The finely tuned -> The finetuned

File: en/docs/basics_topics/optimizer_in_function_config.md [hunk 17]
```diff
@@ -89,13 +39,17 @@ def get_eval_config():
...
+  # build up neural network here
 ```
 
 ### Configuration for training
 
-Same as above, we decorate the `train_job()` function with `@flow.global_function` then we get a job function for training. We can set the optimizer, learning rate and hyperparameters inside the job function. 
+If you specify the `type` parameter of `@flow.global_function` to be `train`, you can get a job function for training.
+
+In the following code, `train_job` is the job function used for training and it is configured with the default `function_config` (so no parameter is passed to `function_config`).
+
```
Comment (MARD1NO) [inline en/docs/basics_topics/optimizer_in_function_config.md:51]: so no parameter is passed to `function_config` -> so there is no parameter passed to `function_config`

File: en/docs/basics_topics/optimizer_in_function_config.md [hunk 18]
```diff
@@ -89,13 +39,17 @@ def get_eval_config():
...
+Because OneFlow will back propagates for `train` functions. Thus you need to specify the following settings like optimizer, learning rate and other hyperparameters in the job function.
```
Comment (MARD1NO) [inline en/docs/basics_topics/optimizer_in_function_config.md:52]: will back propagates -> will back propagate

Comment (MARD1NO) [reply en/docs/basics_topics/optimizer_in_function_config.md:52]: Because 和 Thus 不能连用

File: en/docs/extended_topics/watch_watch_diff.md [hunk 19]
```diff
@@ -1,61 +1,40 @@
...
+## Using Guidance
 
 To get data or gradient tensor in job function, we need to follow these steps:
 
 * Write a callback function and the parameters of the callback function should be annotated to indicate the data type. The logic of the callback function need to be set up by user themselves.
 
 * When defining the job functions, we use `oneflow.watch` or `oneflow.watch_diff` to register callback function. We obtain data tensor from the former one and their corresponding gradient from the latter one.
 
-* OneFlow will call the callback function at an appropriate time at runtime.
+* At the appropriate time when the job function is running, the OneFlow will call the previous callback which was registered earlier and pass the monitored data to the callback function then execute the logic in the callback function.
```
Comment (MARD1NO) [inline en/docs/extended_topics/watch_watch_diff.md:13]: OneFlow 前面的 the 去掉

Comment (MARD1NO) [reply en/docs/extended_topics/watch_watch_diff.md:13]: previous callback 后面加个 function

File: en/docs/quick_start/lenet_mnist.md [hunk 20]
```diff
@@ -71,14 +71,14 @@ python lenet_test.py ./9.png
...
+## Introduction of MNIST Dataset 
 
 MNIST is a handwritten digits database including training set and testing set. Training set includes 60000 pictures and their corresponding label. Yann LeCun and others have normalized all the images and packed them into a single binary file for downloading. http://yann.lecun.com/exdb/mnist/
 
 
-## Define training model
+## Define Training Model
 
-Modules `oneflow.nn` and `oneflow.layers` provide the operators to construct the model.
+Modules  [oneflow.nn](https://oneflow.readthedocs.io/en/master/nn.html) and [oneflow.layers](https://oneflow.readthedocs.io/en/master/layers.html)  provide the operators to construct the model.
```
Comment (MARD1NO) [inline en/docs/quick_start/lenet_mnist.md:81]: [oneflow.nn] 前面多了空格, 类似这些需要自己注意

File: en/docs/quick_start/lenet_mnist.md [hunk 21]
```diff
@@ -123,29 +123,27 @@ def lenet(data, train=False):
 
 As the code showing above, we build up a LeNet network model.
 
-## Implementation of job function for training
+## Implementation of Job Function for Training
 
-OneFlow provides a decorator named `oneflow.global_function` by which we can covert a Python function to a OneFlow **Job Function** .
+OneFlow provides a decorator named  [oneflow.global_function](https://oneflow.readthedocs.io/en/master/oneflow.html#oneflow.global_function)  by which we can covert a Python function to a OneFlow **Job Function** .
```
Comment (MARD1NO) [inline en/docs/quick_start/lenet_mnist.md:128]: 同样的问题,后面也有一些多余的空格

**Review by MARD1NO — CHANGES_REQUESTED** [review]

File: en/docs/basics_topics/build_nn_with_op_and_layer.md [hunk 22]
```diff
@@ -53,16 +49,17 @@ def lenet(data, train=False):
         hidden = flow.nn.dropout(hidden, rate=0.5, name="dropout")
     return flow.layers.dense(hidden, 10, kernel_initializer=initializer, name="dense2")
 ```
+When the job function is run, `data` is in shape of `100x1×28×28`. `data` is firstly used as input in `conv2d` to participate in the convolution calculation and obtained the result `conv1` then `conv1` is passed to `max_pool2d` as input.
```
[the remark this answers is not in the document]
Comment (Ziheng-ZHAO) [reply en/docs/basics_topics/build_nn_with_op_and_layer.md]: `Data` is firstly used as input in `conv2d` to participate in the convolution calculation and  the result from `conv1` is passed to `max_pool2d` as input. 这么说可以吗

**Review by doombeaker — APPROVED** [review]
